osgar/drivers/kloubak.py

```diff
@@ -0,0 +1,283 @@
+        self.last_encoders_time = None
+        self.last_join_angle = None
+
+        self.verbose = False  # should be in Node
+        self.enc_debug_arr = []
+        self.join_debug_arr = []
+        self.count = [0, 0, 0, 0]
+        self.count_arr = []
+
+    def send_pose(self):
+        x, y, heading = self.pose
+        self.publish('pose2d', [round(x*1000), round(y*1000),
+                                round(math.degrees(heading)*100)])
+
+    def update_buttons(self, data):
+        assert len(data) == 1, len(data)
+        val = data[0]
+        if self.buttons is None or val != self.buttons:
+            self.buttons = val
+            stop_status = self.buttons & 0x01 == 0x01
+            if self.emergency_stop != stop_status:
+                self.emergency_stop = stop_status
+                self.bus.publish('emergency_stop', self.emergency_stop)
+                print('Emergency STOP:', self.emergency_stop)
+
+    def update_encoders(self, msg_id, data):
+        assert len(data) == 8, data
+        rpm3, current, duty_cycle = struct.unpack('>ihh', data)
+        if msg_id == CAN_ID_VESC_FRONT_L:
+            self.last_encoders_front_left = rpm3
+#            print('left', rpm3, current, duty_cycle)
+        elif msg_id == CAN_ID_VESC_FRONT_R:
+            self.last_encoders_front_right = rpm3
+#            print('right', rpm3, current, duty_cycle)
+        if msg_id == CAN_ID_VESC_REAR_L:
+            self.last_encoders_rear_left = rpm3
+        elif msg_id == CAN_ID_VESC_REAR_R:
+            self.last_encoders_rear_right = rpm3
+        self.count[msg_id - 0x91] += 1
+#        print(self.count)
+        if self.verbose:
+            self.count_arr.append([self.time.total_seconds()] + self.count)
+
+    def update_pose(self):
+        """Update internal pose with 'dt' step"""
+        if self.last_encoders_front_left is None or self.last_encoders_front_right is None:
+            return False
+        x, y, heading = self.pose
+
+        metricL = ENC_SCALE * self.last_encoders_front_left  # dt is already part of ENC_SCALE
+        metricR = ENC_SCALE * self.last_encoders_front_right
+
+        dist = (metricL + metricR)/2.0
+        angle = (metricR - metricL)/WHEEL_DISTANCE
+
+        # advance robot by given distance and angle
+        if abs(angle) < 0.0000001:  # EPS
+            # Straight movement - a special case
+            x += dist * math.cos(heading)
+            y += dist * math.sin(heading)
+            #Not needed: heading += angle
+        else:
+            # Arc
+            r = dist / angle
+            x += -r * math.sin(heading) + r * math.sin(heading + angle)
+            y += +r * math.cos(heading) - r * math.cos(heading + angle)
+            heading += angle # not normalized
+        self.pose = (x, y, heading)
+        return True
+
+    def process_packet(self, packet, verbose=False):
+        if len(packet) >= 2:
+            msg_id = ((packet[0]) << 3) | (((packet[1]) >> 5) & 0x1f)
```

Comment:
We have this exact same line in eduro.py and spider.py. Maybe it is time for refactoring? Or would you rather keep copy pasting?

Reply:
good point! I would like to "solve" it with changing the structure how CAN messages are sent. Instead of raw couple of bytes it will be (msg_id, data, flags), so this function would no longer be necessary. This CAN refactoring I would do after working pcan on linux (APU) in separate PR - so yes, this should away (later).

Reply:
👍